[13.0][MIG] mass_mailing_list_dynamic: Migration to 13.0

jaredkipe opened the pull request:
Requires #498

There is an issue in the tests for the @api.constrains (around line 82 in tests files)
I can log the `mailing.contact` before entering the constrains and during the constrains and the `list_ids` and `subscription_list_ids` are present before entering the method, and empty in the method itself.  I've flushed various things (flushing on the list is needed now as the count of subscribers has no @api.depends to invalidate the cache) and haven't been able to figure it out, any advice would be helpful.

yajo committed:
[ADD] mass_mailing_list_dynamic: Autopopulated mailing lists

pedrobaeza committed:
[FIX+IMP] mass_mailing_list_dynamic: tests, icons, filters... (#211)

* [FIX+IMP] mass_mailing_list_dynamic: tests, icons, filters...

* Brand new icon
* Added feature of loading an existing filter as criteria
* Tests as SavepointCase for optimizing times
* Tests in post-install for avoiding errors on res.partner not null constraints
  when several modules added them.
* Updated documentation.
* Fix mock in test for not commiting test data.

* [FIX] mass_mailing_list_dynamic: Wasn't able to create contacts in fully synced lists

Syncing context was being set in the wrong object. Added to test too.

* [FIX] mass_mailing_list_dynamic: Allow to write back vals from res.partner

Module mass_mailing_partner writes back certain values from partner to
mass_mailing_contact. Module should allow that write operation.

chienandalu committed:
[FIX] mass_mailing_list_dynamic: Reversible dynamic list

chienandalu committed:
[IMP] mass_mailing_list_dynamic: make user aware of list change

- Adds is_synced field to track whether a dynamic list has unsynced
changes or not so the user is aware that the definitive number of
contacts is yet to be determined.
- It fixes an issue that made impossible deleting a res.partner filter
when a list had use it to filter contacts.
- It also shows only the filters available for the user (shared and
belonging to self).

Commit:
[MIG] mass_mailing_list_dynamic: Migration to 12.0

ernestotejeda committed:
[MIG] mass_mailing_list_dynamic: Fix contact synchronization

jaredkipe committed:
[IMP] mass_mailing_list_dynamic: black, isort

jaredkipe committed:
[MIG] mass_mailing_list_dynamic: Migration to 13.0

pedrobaeza commented:
@ernestotejeda @manuelcalerosolis please review this TT22860

pedrobaeza commented:
@jaredkipe please rebase after merging the dependency

jaredkipe commented:
@pedrobaeza all done with rebase

Any ideas on testing the @api.constrains ?

pedrobaeza commented:
I haven't checked. My colleagues may say when reviewing.

ernestotejeda commented:
Hello @jaredkipe , it has to do with this https://github.com/OCA/social/blob/5c87e6b7e2b6e566ed09e0c688cce8a719faec22/mass_mailing_partner/models/mailing_contact.py#L67. Take a look at the redefinition of the create method too that is in that same file a few lines above. Maybe those two method should be reviewed again.

jaredkipe commented:
Thanks that is a really great observation!

I was concerned with the constraint in this module and didn't think to look up into dependencies for something like this... probably because I thought the constraint would trigger off of write itself, which of course in this case doesn't happen (the super/base implementation) until the new values are already calculated.

jaredkipe committed:
[FIX] mass_mailing_partner: Unsubscribe should not unlink the subscription.

Before, on the backend if you check the `opt_out` checkbox and saving on a `mailing.contact`, the entire row (`mailing.contact.subscription`) would be unlinked.  Combined with `mass_mailing_list_dynamic` this means the contact would just be re-subscribed over and over.

After, check `opt_out` will fill `unsubscription_date` and the row will stay.  Tests in `mass_mailing_list_dynamic` pass.

yajo commented on mass_mailing_list_dynamic/models/mail_mass_mailing_contact.py:
```diff
@@ -9,13 +9,15 @@
+                raise ValidationError(
+                    _("""Cannot edit manually contacts in a fully synchronized list.
+                        Change its sync method or execute
+                        a manual sync instead."""))
```
If you don't touch this string, we will not have to translate it again.

jaredkipe replied:
I'm not editing things for giggles, if this changed it would be from running `black` or other linting/clean up as prescribed by the documentation.

jaredkipe replied:
Probably was an error on my part, however this file was renamed to  `mailing_contact.py` and does not have this problem.

simahawk commented:
@jaredkipe any chance to move fwd?

chienandalu committed:
[FIX] mass_mailing_list_dynamic: list sync on send

- We change the method to hook on (it wasn't correctly overriden anyway)
to ensure the list is resynced even if no recipients are left. For
example: we create a dynamic list with a domain expecting it to sync in
the first mass_mailing. It wouldn't sync as Odoo dismisses the mailing
considering no recipients are left thus it wouldn't get into the
`send_mail` method.

rafaelbn commented:
Hello @jaredkipe ,

Please could you update this PR including https://github.com/OCA/social/pull/571

Thank you!!

cc @chienandalu

jaredkipe commented:
@rafaelbn Rebased, cherry pick wasn't clean because of other changes, including that this method is renamed in 13

chienandalu commented on mass_mailing_list_dynamic/models/mailing.py:
```diff
@@ -1,4 +1,5 @@
 # Copyright 2017 Tecnativa - Jairo Llopis
+# Copyright 2020 Hibou Corp. - Jared Kipe
```
Why do you author this change and commit? :thinking:

chienandalu reviewed (changes requested):
Works as expected. Anyway, you should fix the commit history respecting authorship, please

jaredkipe commented:
I have re-done the history to put the [FIX] #571 , with the correct author before the migration.

However, it came to my attention after I wrote this migration that I should be adding myself to the copyright on materially changed files. I understand why it looked hinky, so I've done that in the migration commit now.

Also, it worked "as expected" before yesterday.  There was no reason that I was aware of that this wouldn't have been merged before the sudden need to put a 6 hour old fix into a 4 month old migration, which would have been a smoother process for everyone.

@chienandalu would you mind checking to see if 13.0 `mass_mailing_partner` works as expected, and consider merging this 4 month old [FIX] #523

chienandalu commented:
> I have re-done the history to put the [FIX] #571 , with the correct author before the migration.

Thanks :)

> However, it came to my attention after I wrote this migration that I should be adding myself to the copyright on materially changed files. I understand why it looked hinky, so I've done that in the migration commit now.

Well, not really. The ettiquette would be the following:

- Standard migration tweaks (like renames of fields, models, views...) don't qualify for authorship
- Typos or translations: contributor
- Tests adaptations to the new version: depending on the entity of the change.
- Functional changes to adapt to new version: qualify you as author.
- Bug fixes: qualify you as author

Of course, for everyone of these points, you're considered a contributor and you can put your credentials in the CONTRIBUTORS.rst

> Also, it worked "as expected" before yesterday. There was no reason that I was aware of that this wouldn't have been merged before the sudden need to put a 6 hour old fix into a 4 month old migration, which would have been a smoother process for everyone.

Well, the thing is it didn't, as it's missing a fix that was detected and fixed 6 hours ago :slightly_smiling_face: 

> @chienandalu would you mind checking to see if 13.0 `mass_mailing_partner` works as expected, and consider merging this 4 month old [FIX] #523

ping me there and I check it

pedrobaeza commented:
Correction: bug fixes don't usually qualify for co-authorship.

yajo commented on mass_mailing_list_dynamic/wizards/mailing_load_filter.py:
```diff
@@ -1,12 +1,13 @@
-    _name = "mail.mass_mailing.load.filter"
-    _description = "Mail Mass Mailing Load Filter"
+class MailingLoadFilter(models.TransientModel):
+    _name = "mailing.load.filter"
```
@pedrobaeza Please correct me if I'm wrong: this is OK and it doesn't need a migration script because it is a transient model and there's no problem if the previous table gets dumped and a new one is created. Right?

pedrobaeza replied:
Yes, transient models are not mandatory to be renamed.

pedrobaeza commented:
/ocabot merge nobump
